src/WorkflowFunction.ts

```diff
@@ -0,0 +1,214 @@
+    throw new WorkflowFunctionInitializationError(errorMsg);
+  }
+
+  // Ensure all listeners are functions
+  listeners.forEach((listener) => {
+    if (!(listener instanceof Function)) {
+      const errorMsg = 'All WorkflowFunction listeners must be functions';
+      throw new WorkflowFunctionInitializationError(errorMsg);
+    }
+  });
+}
+
+/**
+ * `processStepMiddleware()` invokes each callback for lifecycle event
+ * @param args workflow_step_edit action
+ */
+export async function processStepMiddleware(
+  args: AllWorkflowFunctionMiddlewareArgs,
+  middleware: WorkflowFunctionMiddleware,
+): Promise<void> {
+  const { context, client, logger } = args;
+  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
+  const lastCallback = callbacks.pop();
+
+  if (lastCallback !== undefined) {
+    await processMiddleware(
+      callbacks, args, context, client, logger,
+      async () => lastCallback({ ...args, context, client, logger }),
+    );
+  }
+}
+
+export function isFunctionEvent(args: AnyMiddlewareArgs): args is AllWorkflowFunctionMiddlewareArgs {
+  return VALID_PAYLOAD_TYPES.has(args.payload.type);
+}
+
+function selectToken(context: Context): string | undefined {
+  // If attachFunctionToken = false, fallback to botToken or userToken
+  return context.functionBotToken ? context.functionBotToken : context.botToken || context.userToken;
+}
+
+/**
+ * Factory for `complete()` utility
+ * @param args function_executed event
+ */
+export function createFunctionComplete(
+  args: AllWorkflowFunctionMiddlewareArgs<WorkflowFunctionExecuteMiddlewareArgs>,
+): FunctionCompleteFn {
+  const {
+    context,
+    client,
+    payload: { function_execution_id },
+  } = args;
+  const token = selectToken(context);
+
+  return (params: Parameters<FunctionCompleteFn>[0] = {}) => client.functions.completeSuccess({
+    token,
+    outputs: params.outputs || {},
+    function_execution_id,
+  });
+}
+
+/**
+ * Factory for `fail()` utility
+ * @param args function_executed event
+ */
+export function createFunctionFail(
+  args: AllWorkflowFunctionMiddlewareArgs<WorkflowFunctionExecuteMiddlewareArgs>,
+): FunctionFailFn {
+  const {
+    context,
+    client,
+    payload: { function_execution_id },
+  } = args;
+  const token = selectToken(context);
+
+  return (params: Parameters<FunctionFailFn>[0]) => {
+    const { error } = params ?? {};
+
+    return client.functions.completeError({
+      token,
+      error,
+      function_execution_id,
+    });
+  };
+}
+
+/**
+ * `prepareFunctionArgs()` takes in a function's args and:
+ *  1. removes the next() passed in from App-level middleware processing
+ *    - events will *not* continue down global middleware chain to subsequent listeners
+ *  2. augments args with step lifecycle-specific properties/utilities
+ * */
+export function prepareFunctionArgs(args: any): AllWorkflowFunctionMiddlewareArgs {
```

Comment:
🤓 Naming nit!

In the deno SDK we call this process ["enriching" the context](https://github.com/slackapi/deno-slack-sdk/blob/5413c437e61f661e85a947eb3995c4c4870d7e57/src/functions/enrich-context.ts#L7). Can we converge on the same terminology?

Reply:
Carryover from the existing `WorkflowStep` architecture and was done this way to align with the way we refer to 'args' elsewhere in listeners/middleware in `App.ts`. If you feel strongly about it, I'm fine with changing it! 🙂

Reply:
(plus, we have an actual `context` object handled in `App.ts` that is different than what is done here)

Reply:
OK I leave the naming to you. Anything in software that uses the word 'prepare' to me is a pet peeve as I find it doesn't communicate anything intrinsically - I always need to read the source to understand what the function does.

Reply:
Ah, I see your meaning. No issue with swapping out `prepare` for `enrich`.